persist/fs/files.go

```diff
@@ -324,20 +460,91 @@ func findSubDirectoriesAndPaths(directoryPath string) (directoryNamesToPaths, er
 	return subDirectoriesToPaths, nil
 }
 
-func filesetFiles(filePathPrefix string, namespace ident.ID, shard uint32, pattern string) ([]string, error) {
-	shardDir := ShardDirPath(filePathPrefix, namespace, shard)
-	return findFiles(shardDir, pattern, func(files []string) sort.Interface {
+func snapshotFiles(filePathPrefix string, namespace ident.ID, shard uint32, pattern string) ([]SnapshotFile, error) {
+	shardDir := ShardSnapshotsDirPath(filePathPrefix, namespace, shard)
+	byTimeAsc, err := findFiles(shardDir, pattern, func(files []string) sort.Interface {
+		return snapshotsByTimeAndIndexAscending(files)
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		latestBlockStart   time.Time
+		latestIndex        int
+		latestSnapshotFile SnapshotFile
+		snapshotFiles      = []SnapshotFile{}
+	)
+	for _, file := range byTimeAsc {
+		currentFileBlockStart, index, err := TimeAndIndexFromSnapshotFilename(file)
+		if err != nil {
+			return nil, err
+		}
+
+		if latestBlockStart.IsZero() {
+			latestSnapshotFile = NewSnapshotFile(currentFileBlockStart)
+			latestIndex = index
+		} else if !currentFileBlockStart.Equal(latestBlockStart) || latestIndex != index {
+			snapshotFiles = append(snapshotFiles, latestSnapshotFile)
+			latestSnapshotFile = NewSnapshotFile(currentFileBlockStart)
+			latestIndex = index
+		}
+		latestBlockStart = currentFileBlockStart
+
+		latestSnapshotFile.Index = index
+		latestSnapshotFile.Files = append(latestSnapshotFile.Files, file)
+	}
+	snapshotFiles = append(snapshotFiles, latestSnapshotFile)
+
+	return snapshotFiles, nil
+}
+
+func filesetFiles(filePathPrefix string, namespace ident.ID, shard uint32, pattern string) (FilesetFilesSlice, error) {
+	shardDir := ShardDataDirPath(filePathPrefix, namespace, shard)
+	byTimeAsc, err := findFiles(shardDir, pattern, func(files []string) sort.Interface {
 		return byTimeAscending(files)
 	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		latestBlockStart  time.Time
+		latestFilesetFile FilesetFile
+		filesetFiles      = []FilesetFile{}
+	)
+	for _, file := range byTimeAsc {
+		currentFileBlockStart, err := TimeFromFileName(file)
+		if err != nil {
+			return nil, err
+		}
+
+		if latestBlockStart.IsZero() {
+			latestFilesetFile = NewFilesetFile(currentFileBlockStart)
+		} else if !currentFileBlockStart.Equal(latestBlockStart) {
```

Comment:
Will this happen for fileset files as well as snapshots?

Reply:
Yeah the way the code works is that it basically just asks the O.S for any files that look like a fileset file (checkpoints, data files, index files, info files etc etc) and they all come back with no particular grouping. So I sorted them by time at the top of this method, and then I "step" through them grouping them together into block starts. The condition you've highlighted signals that we've reached the end of a group of file and are moving on to the next block start.

Its like a reduce basically

Reply:
Oh I see, I thought this was the index number grouping.